---
layout: post
title: "Petites astuces avec JMX"
date: 2011-01-23
comments: false
categories:
 - java
 - jmx
---

<div class='post'>
<div class="separator" style="clear: both; text-align: center;"><a href="http://1.bp.blogspot.com/_XLL8sJPQ97g/TTr8YAhjHqI/AAAAAAAAAUA/nx8ug8MDDNA/s1600/jmx02.png" imageanchor="1" style="clear: left; float: left; margin-bottom: 1em; margin-right: 1em; text-align: justify;"><img border="0" height="226" src="http://1.bp.blogspot.com/_XLL8sJPQ97g/TTr8YAhjHqI/AAAAAAAAAUA/nx8ug8MDDNA/s320/jmx02.png" width="320" /></a></div><div style="text-align: justify;">Ce rapide article fournit un petit retour d'expérience de quelques galères que j'ai pu avoir avec JMX (<i>Java Management eXtension</i>), galères certes stupides mais qui pourront peut être servir à d'autres... ;-)<br />Sera donc abordé deux points :<br /><br /><ol><li>Erreur lors de l'arrêt de tomcat configuré avec JMX</li><li>Impossibilité de se connecter à l'agent JMX</li></ol></div><!-- more --><div style="text-align: justify;"><br /></div><div style="text-align: justify;"><br /><br /></div><div style="text-align: justify;"><br /></div><h1>Message d'erreur lors de l'arrêt d'Apache Tomcat?</h1><h2>Cas d'usage</h2><div style="text-align: justify;">Les options permettant d'activer JMX sur la JVM ont été passées via la variable JAVA_OPTS ou autre (via le catalina.sh par exemple).</div><div style="text-align: justify;">Le démarrage de Tomcat s'effectue sans problème mais lors de son arrêt, une erreur&nbsp;apparaît&nbsp;et tomcat refuse de s'arrêter c'est-à-dire que son processus tourne toujours en tâche de fond.</div><div style="text-align: justify;"><br /></div><div style="text-align: justify;">Exemple de variable JAVA_OPTS :</div><pre class="brush: bash">export JAVA_OPTS="$JAVA_OPTS \<br />-Dcom.sun.management.jmxremote \<br />-Dcom.sun.management.jmxremote.port=18080 \<br />-Dcom.sun.management.jmxremote.authenticate=false \<br />-Dcom.sun.management.jmxremote.ssl=false"</pre><div style="text-align: justify;"><br /></div><div style="text-align: justify;">Message affiché sur la console lors du démarrage de Tomcat :</div><pre class="brush: bash">khanh@khanh-vaio:/opt/apache-tomcat-6.0.20/bin$ ./startup.sh<br />Using CATALINA_BASE:   /opt/apache-tomcat-6.0.20<br />Using CATALINA_TMPDIR:  /opt/apache-tomcat-6.0.20<br />Using CATALINA_HOME:  /opt/apache-tomcat-6.0.20/temp<br />Using JRE_HOME:     /usr/lib/jvm/java-6-sun/jre</pre><div style="text-align: justify;"><br /></div><div style="text-align: justify;">Message affiché sur la console lors de l'arrêt de Tomcat :</div><pre class="brush: bash">khanh@khanh-vaio:/opt/apache-tomcat-6.0.20/bin$ ./shutdown.sh<br />Using CATALINA_BASE:   /opt/apache-tomcat-6.0.20<br />Using CATALINA_TMPDIR:  /opt/apache-tomcat-6.0.20<br />Using CATALINA_HOME:   /opt/apache-tomcat-6.0.20/temp<br />Using JRE_HOME:     /usr/lib/jvm/java-6-sun/jre<br />Erreur: Exception envoyée par l'agent : java.rmi.server.ExportException: Port already in use: 18080; nested exception is: <br />        java.net.BindException: Address already in use<br /></pre><h2>Explication</h2><div style="text-align: justify;">En fait, cette erreur se produit lors de l'arrêt de Tomcat car le script shutdown.sh (ou catalina.sh stop) relance un processus java qui tente alors de démarrer un serveur RMI qui est utilisé par JMX. Bien sûr, ce port étant déjà utilisé par le processus java que l'on tente d'arrêter, cela échoue.</div><h2>Proposition pour résoudre le problème</h2><div style="margin-bottom: 0px; margin-left: 0px; margin-right: 0px; margin-top: 0px;"><ul><li style="text-align: justify;"><s>Ne pas positionner les options pour activer JMX sur la variable système JAVA_OPTS au risque de ne pas pouvoir démarrer plusieurs processus Java.</s></li><li style="text-align: justify;"><s>Si c'est le script catalina.sh qui positionne la variable JAVA_OPTS, ne pas la positionner dans tous les cas mais seulement dans le cas où l'argument start est passé au script.</s></li><li style="text-align: justify;"><s>Positionner la variable JAVA_OPTS dans le script startup.sh.</s></li><li style="text-align: justify;">Utiliser la variable CATALINA_OPTS qui n'est utilisée que quand les options run et start sont passées"</li></ul><blockquote><blockquote>CATALINA_OPTS &nbsp; (Optional) Java runtime options used when the "start", &nbsp;or "run" command is executed.</blockquote></blockquote></div><h1>Problème de connexion?</h1><h2>Cas d'usage</h2><div style="text-align: justify;">Une jconsole ou une jvisualvm qui n'arrive pas à se connecter à un programme ou une connexion au connecteur d'un agent JMX qui échoue avec une exception.</div><div style="text-align: justify;">Chose d'autant plus étonnante qu'avec une connexion réseau, il n'y a pas de problèmes alors qu'en mode hors-ligne, impossible d'avoir accès à l'agent JMX.</div><div style="text-align: justify;"><br /></div><div style="text-align: justify;">Exemple d'exception lors de la tentative de connexion à l'agent JMX distant :</div><pre class="brush: bash">java.lang.NullPointerException<br />at javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:281)<br />at javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:227)<br />at com.blogspot.jetoile.jgroups.connector.ConnectorStubListener.update(ConnectorStubListener.java:47)<br />at com.blogspot.jetoile.jgroups.connector.ConnectorStubManager.put(ConnectorStubManager.java:46)<br />[...]<br />at com.blogspot.jetoile.jgroups.connector.TestJMXServerTest.main(TestJMXServerTest.java:34)<br /></pre><div style="text-align: justify;"><br /></div><div style="text-align: justify;">Fenêtre d'erreur de la jconsole :</div><div class="separator" style="clear: both; text-align: center;"><a href="http://3.bp.blogspot.com/_XLL8sJPQ97g/TTsrkOibDDI/AAAAAAAAAUE/qpX-1nw0OOA/s1600/error02.png" imageanchor="1" style="margin-left: 1em; margin-right: 1em;"><img border="0" height="333" src="http://3.bp.blogspot.com/_XLL8sJPQ97g/TTsrkOibDDI/AAAAAAAAAUE/qpX-1nw0OOA/s400/error02.png" width="400" /></a></div><div style="text-align: justify;">Fenêtre d'erreur de jvisualvm : "<i>Data not available because JMX connection to the JMX agent could not be established.</i>" :</div><div class="separator" style="clear: both; text-align: center;"><a href="http://4.bp.blogspot.com/_XLL8sJPQ97g/TTsroeTxKoI/AAAAAAAAAUI/0aUghCdjv0w/s1600/error01.png" imageanchor="1" style="margin-left: 1em; margin-right: 1em;"><img border="0" height="243" src="http://4.bp.blogspot.com/_XLL8sJPQ97g/TTsroeTxKoI/AAAAAAAAAUI/0aUghCdjv0w/s400/error01.png" width="400" /></a></div><h2>Analyse</h2><div><div style="text-align: justify;">Après de multiples tentatives pour comprendre, j'ai tenté un changement de jre pour voir si cela ne pouvait pas venir d'une anomalie (si j'avais un problème de connexion en mode hors-ligne, ce n'était pas pour rien puisque j'étais effectivement hors-ligne et, par conséquent, je devais me débrouiller par moi-même... ;-) ).</div><div style="text-align: justify;">Ainsi, en passant d'une jre hotspot vers une openjdk, j'ai obtenu une stacktrace beaucoup plus complète puisqu'elle m'a clairement indiquée la source du problème :</div><br /><pre class="brush: bash">ERROR c.b.j.j.c.ConnectorStubListener - ConnectorStubListener update: {}<br />java.rmi.ConnectIOException: Exception creating connection to: 192.168.xxx.xxx; nested exception is:<br />java.net.NoRouteToHostException: Network is unreachable<br />at sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:632) ~[na:1.6.0_20]<br />at sun.rmi.transport.tcp.TCPChannel.createConnection(TCPChannel.java:216) ~[na:1.6.0_20]<br />at sun.rmi.transport.tcp.TCPChannel.newConnection(TCPChannel.java:202) ~[na:1.6.0_20]<br />at sun.rmi.server.UnicastRef.invoke(UnicastRef.java:128) ~[na:1.6.0_20]<br />at javax.management.remote.rmi.RMIServerImpl_Stub.newClient(Unknown Source) ~[na:1.6.0_20]<br />at javax.management.remote.rmi.RMIConnector.getConnection(RMIConnector.java:2343) ~[na:1.6.0_20]<br />at javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:296) ~[na:1.6.0_20]<br />at javax.management.remote.rmi.RMIConnector.connect(RMIConnector.java:244) ~[na:1.6.0_20]<br />at com.blogspot.jetoile.jgroups.connector.ConnectorStubListener.update(ConnectorStubListener.java:47) ~[classes/:na]<br />at com.blogspot.jetoile.jgroups.connector.ConnectorStubManager.put(ConnectorStubManager.java:46) [classes/:na]<br />...<br />at com.blogspot.jetoile.jgroups.connector.JMXServer.start(JMXServer.java:152) [classes/:na]<br />at com.blogspot.jetoile.jgroups.connector.TestJMXServerTest2.main(TestJMXServerTest2.java:34) [test-classes/:na]<br /><br /><br />Caused by: java.net.NoRouteToHostException: Network is unreachable<br />at java.net.PlainSocketImpl.socketConnect(Native Method) ~[na:1.6.0_20]<br />at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:310) ~[na:1.6.0_20]<br />at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:176) ~[na:1.6.0_20]<br />at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:163) ~[na:1.6.0_20]<br />at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:384) ~[na:1.6.0_20]<br />at java.net.Socket.connect(Socket.java:546) ~[na:1.6.0_20]<br />at java.net.Socket.connect(Socket.java:495) ~[na:1.6.0_20]<br />at java.net.Socket.<init>(Socket.java:392) ~[na:1.6.0_20]</init><br />at java.net.Socket.<init>(Socket.java:206) ~[na:1.6.0_20]</init><br />at sun.rmi.transport.proxy.RMIDirectSocketFactory.createSocket(RMIDirectSocketFactory.java:40) ~[na:1.6.0_20]<br />at sun.rmi.transport.proxy.RMIMasterSocketFactory.createSocket(RMIMasterSocketFactory.java:146) ~[na:1.6.0_20]<br />at sun.rmi.transport.tcp.TCPEndpoint.newSocket(TCPEndpoint.java:613) ~[na:1.6.0_20]<br />... 34 common frames omitted<br /></pre></div><h2>Explication</h2><div style="text-align: justify;">En fait, mon problème venait du fait que mon fichier <i>hosts</i> était incorrect (je vous avais prévenu que c'était une erreur stupide : n'importe qui aurait pensé à vérifier sa configuration réseau... ! ;-) ).&nbsp;</div><div style="text-align: justify;">En effet, la couche logicielle gérant mes connexions réseaux sur mon ordinateur a tendance à ajouter dans mon fichier /etc/hosts mon adresse réseau. Cependant, lorsque je coupe mon wifi, mon ordinateur passe en mode hors-ligne sans modifier mon fichier <i>hosts</i>, ce qui corrompt ma configuration réseau.</div><h2>Proposition pour résoudre le problème</h2><div><ul><li style="text-align: justify;">Corriger son fichier <i>hosts</i> et vérifier sa configuration réseau</li></ul><div style="text-align: justify;">A noter que dans mon cas, le fait d'avoir changer de jre m'a permis d'obtenir une stacktrace beaucoup plus complète et précise. Aussi, dans un moment de désespoir et en dernier recours, il peut être intéressant de changer de jre ;-).</div></div></div>
<h2>Comments</h2>
<div class='comments'>
<div class='comment'>
<div class='author'>Khanh</div>
<div class='content'>
En effet, d&#39;ailleurs il s&#39;agit même de la seule bonne réponse (Comme quoi, lire la doc peut servir...)! Merci de m&#39;y avoir fait penser ;-) Je rajoute cette possibilité.</div>
</div>
<div class='comment'>
<div class='author'>Anonymous</div>
<div class='content'>
Pour le problème 1 : <br />une solution est aussi d&#39;utiliser la variable d&#39;environnement CATALINA_OPTS</div>
</div>
</div>
